File: sdks/python/src/ice/context/context.py
"""Context implementation for Ice SDK."""

from __future__ import annotations

import io
import time
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ice.context.pack import Pack


class Context:
    """
    Context is the execution context for ice processing.
    
    Contains:
    - ice_time: Time when context was created
    - ice_id: Handler ID being executed
    - pack: Input pack
    - process_info: Debug information collected during execution
    """
    
    def __init__(self, ice_id: int, pack: Pack) -> None:
        from ice.context.pack import Pack as PackClass
        
        self.ice_time: int = int(time.time() * 1000)
        self.ice_id: int = ice_id
        self.pack: Pack = pack if pack is not None else PackClass()
        self.process_info: io.StringIO = io.StringIO()
    
    def get_process_info(self) -> str:
        """Get collected process information."""
        return self.process_info.getvalue()
    
    def __repr__(self) -> str:
        return f"Context(ice_id={self.ice_id}, pack={self.pack})"

File: sdks/go/context/context.go
package context

import (
	"strings"
	"sync/atomic"
	"time"
)

// Context represents the execution context for ice processing.
type Context struct {
	// IceTime is the time when the context was created.
	IceTime int64
	// IceId is the handler ID being executed.
	IceId int64
	// Pack is the input pack.
	Pack *Pack
	// ProcessInfo collects debug information during execution.
	ProcessInfo *strings.Builder
}

// NewContext creates a new Context.
func NewContext(iceId int64, pack *Pack) *Context {
	if pack == nil {
		pack = NewPack()
	}
	return &Context{
		IceTime:     time.Now().UnixMilli(),
		IceId:       iceId,
		Pack:        pack,
		ProcessInfo: &strings.Builder{},
	}
}

// ParallelContext wraps a Context for parallel execution.
type ParallelContext struct {
	isDone atomic.Bool
	Ctx    *Context
}

// NewParallelContext creates a new ParallelContext.
func NewParallelContext(ctx *Context) *ParallelContext {
	return &ParallelContext{
		Ctx: ctx,
	}
}

// IsDone returns whether the parallel execution is done.
func (pc *ParallelContext) IsDone() bool {
	return pc.isDone.Load()
}

// SetDone marks the parallel execution as done.
func (pc *ParallelContext) SetDone() {
	pc.isDone.Store(true)
}
